In [1]:

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_score
import plotly.express as px
import plotly.graph_objects as go




#https://thecleverprogrammer.com/2022/03/01/future-sales-prediction-with-machine-learning/
#https://campus.datacamp.com/courses/supervised-learning-with-scikit-learn/regression-6320c92e-31c3-48fb-9382-6a9169125722?ex=6
# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory
#https://app.datacamp.com/learn/courses/introduction-to-regression-with-statsmodels-in-python
import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# Predicción de ventas futuras (estudio de caso)
El conjunto de datos contiene los datos sobre las ventas de  productos. Ademas el conjunto de datos trata sobre los costos publicitarios incurridos por la empresa en varias plataformas publicitarias. A continuación se muestra la descripción de todas las columnas del conjunto de datos:

* TV:  Costo publicitario gastado en dólares para publicidad en TV;
* Radio:  Costo publicitario gastado en dólares para publicidad en Radio;
* Periódico:  Costo publicitario gastado en dólares para publicidad en Periódico;
* Ventas:  Número de unidades vendidas

Este soy yo ->https://www.linkedin.com/in/dmcarrera/


-------------------------------------------------------------------------------------

Future Sales Prediction (Case Study)
The dataset contains data on product sales. Additionally, the dataset deals with advertising costs incurred by the company on various advertising platforms. Below is the description of all columns in the dataset:

TV: Advertising cost spent in dollars for TV advertising;
Radio: Advertising cost spent in dollars for Radio advertising;
Newspaper: Advertising cost spent in dollars for Newspaper advertising;
Sales: Number of units sold

# Exploracion de datos

Resumen de estadísticas descriptivas: Calculo medidas estadísticas como la media, la mediana, la desviación estándar, el mínimo y el máximo para cada variable.

Dimensiones del conjunto de datos: Examino la cantidad de filas y columnas en el conjunto de datos.

Valores Nulos: Identifico valores nulos que pueden traer inconvenientes a la hora de modelar

----------------------------------------------------------------------------
Data Exploration
Summary of Descriptive Statistics: I calculate statistical measures such as the mean, median, standard deviation, minimum, and maximum for each variable.

Dataset Dimensions: I examine the number of rows and columns in the dataset.

Missing Values: I identify null values that may pose challenges when modeling.

In [2]:
data = pd.read_csv('https://raw.githubusercontent.com/amankharwal/Website-data/master/advertising.csv')

print(data.head())
print(data.info())

print(data.describe())

      TV  Radio  Newspaper  Sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3   12.0
3  151.5   41.3       58.5   16.5
4  180.8   10.8       58.4   17.9
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   TV         200 non-null    float64
 1   Radio      200 non-null    float64
 2   Newspaper  200 non-null    float64
 3   Sales      200 non-null    float64
dtypes: float64(4)
memory usage: 6.4 KB
None
               TV       Radio   Newspaper       Sales
count  200.000000  200.000000  200.000000  200.000000
mean   147.042500   23.264000   30.554000   15.130500
std     85.854236   14.846809   21.778621    5.283892
min      0.700000    0.000000    0.300000    1.600000
25%     74.375000    9.975000   12.750000   11.000000
50%    149.750000   22.900000   25.750000   16.000000
75%    218.825000   36.525000

In [3]:
print(data.isnull().sum())

TV           0
Radio        0
Newspaper    0
Sales        0
dtype: int64


# Visualizacion
Exploro las relaciones entre canles publicitarios para entender si existe rellaciones directa o no,  con las ventas

-------------------------------------------------------------------------------
Visualization
I explore the relationships between advertising channels to understand whether there is a direct correlation with sales.

En este grafico veremos la relacion entre publicidad en Television y cantidad de ventas

---------------------------------------------------------------------------------
In this graph, we will see the relationship between television advertising and the quantity of sales

In [4]:
figure = px.scatter(data_frame= data, 
                   x="Sales",
                   y="TV",
                   size="TV", trendline="ols")
figure.show()

En este grafico veremos la relacion entre publicidad en prediodicos y cantidad de ventas

-----------------------------------------------------------------------------

In this graph, we will see the relationship between newspaper advertising and the quantity of sales.

In [5]:
figure = px.scatter(data_frame= data, 
                   x="Sales",
                   y="Newspaper",
                   size="Newspaper", trendline="ols")
figure.show()

En este grafico veremos la relacion entre publicidad en Radio y cantidad de ventas

-------------------------------------------------------------------------------
In this graph, we will see the relationship between radio advertising and the quantity of sales.

In [6]:
figure = px.scatter(data_frame= data, 
                   x="Sales",
                   y="Radio",
                   size="Radio", trendline="ols")
figure.show()

Del total gastado en publicidad en diversas plataformas, puedo afirmar que la inversión en publicidad en televisión genera más ventas del producto. Ahora, analizare la correlación entre todas las columnas y la columna de ventas:

----------------------------------------------------------------------------

Of the total amount spent on advertising across various platforms, I can assert that the investment in television advertising generates more product sales. Now, I will analyze the correlation between all columns and the sales column:

In [7]:
correlation = data.corr()
correlation['Sales'].sort_values(ascending = False)

Sales        1.000000
TV           0.901208
Radio        0.349631
Newspaper    0.157960
Name: Sales, dtype: float64

# Entrenamiento
Ahora, en esta sección, entrenaré un modelo de aprendizaje automático para predecir las ventas futuras de un producto. Pero antes de entrenar el modelo, voy a  dividir  los datos en conjuntos de entrenamiento y de prueba:

-------------------------------------------------------------------------------
Training
Now, in this section, I will train a machine learning model to predict future product sales. But before training the model, I will split the data into training and testing sets:

In [8]:
x = np.array(data.drop(["Sales"],axis = 1))

y = np.array(data["Sales"]) #variable objetivo

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [9]:
model = LinearRegression()
model.fit(X_train, y_train)
print(model.score(X_test, y_test))

0.9059011844150826


# Evaluacion del Modelo

El objetivo de evaluar un modelo de regresión lineal es determinar su capacidad para predecir de manera precisa las respuestas de un fenómeno lineal. Se busca medir la calidad de las predicciones, identificar posibles errores y ajustar el modelo para mejorar su capacidad predictiva en datos no vistos. Esto se logra mediante métricas como el error cuadrático medio y la validación cruzada, garantizando que el modelo sea robusto y generalice bien a nuevas observaciones.

--------------------------------------------------------------------------------------


The goal of evaluating a linear regression model is to assess its ability to accurately predict responses in a linear phenomenon. The objective is to measure the quality of predictions, identify potential errors, and fine-tune the model to enhance its predictive capability on unseen data. This is achieved through metrics such as mean squared error and cross-validation, ensuring that the model is robust and generalizes well to new observations.

In [10]:
print(model.score(X_test, y_test))

0.9059011844150826


# R^2
Descripcion
* R^2 = 0: El modelo no explica ninguna variabilidad en la variable dependiente.
* R^2 = 1: El modelo explica toda la variabilidad en la variable dependiente.

En este  caso, un R^2 del 90.5% sugiere que el modelo está capturando y explicando una proporción significativa de la variabilidad en tus datos. 

------------------------------------------------------------------------

Description

* R^2 = 0: The model explains no variability in the dependent variable.
* R^2 = 1: The model explains all the variability in the dependent variable.

In this case, an R^2 of 90.5% suggests that the model is capturing and explaining a significant proportion of the variability in your data.

In [11]:
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
print(mse, rmse)

2.9077569102710923 1.7052146229349232


# Error Cuadrático Medio (MSE):

Es la media de los errores al cuadrado entre las predicciones y los valores reales.
Cuanto menor sea el MSE, mejor será el rendimiento del modelo.
MSE penaliza fuertemente los errores grandes.
El MSE es aproximadamente 2.91.

# Raíz del Error Cuadrático Medio (RMSE):

Es la raíz cuadrada del MSE y tiene la misma unidad que la variable de respuesta.
Proporciona una interpretación más intuitiva porque está en la misma escala que la variable de respuesta.
Cuanto menor sea el RMSE, mejor será el rendimiento del modelo.
El RMSE es aproximadamente 1.71.

Un RMSE de 1.71 en particular indica que, en promedio, las predicciones están desviadas por alrededor de 1.71 unidades en la misma escala que la variable de respuesta.

----------------------------------------------------------------------------
* Mean Squared Error (MSE):
It is the average of the squared errors between predictions and actual values. The lower the MSE, the better the model's performance. MSE strongly penalizes large errors. The MSE is approximately 2.91.

* Root Mean Squared Error (RMSE):
It is the square root of MSE and has the same unit as the response variable. It provides a more intuitive interpretation because it is on the same scale as the response variable. The lower the RMSE, the better the model's performance. The RMSE is approximately 1.71.

A specific RMSE of 1.71 indicates that, on average, predictions deviate by around 1.71 units on the same scale as the response variable.

In [12]:
residuals = y_test - y_pred
print(residuals.mean())


-0.39850338503085425


# Valor Residual Promedio

Un valor negativo indica que, en promedio, el modelo tiende a subestimar las unidades vendidas en comparación con los valores reales. En este caso, el valor de -0.3985 implica que, en promedio, el modelo predice aproximadamente un 0.4% menos unidades de producto vendidas de lo que realmente se observa.

------------------------------------------------------------------------
Average Residual Value
A negative value indicates that, on average, the model tends to underestimate the units sold compared to the actual values. In this case, the value of -0.3985 implies that, on average, the model predicts approximately 0.4% fewer units of the product sold than what is actually observed.

In [13]:
#coediciente del modelo
print(model.coef_)

[0.05450927 0.10094536 0.00433665]


# Coeficiente:
Interpretacion
Gasto en Publicidad en TV (0.0545):

Por cada dólar adicional gastado en publicidad en TV, se espera un aumento de aproximadamente 0.0545 unidades en las ventas.
Gasto en Publicidad en Radio (0.1009):

Por cada dólar adicional gastado en publicidad en Radio, se espera un aumento de aproximadamente 0.1009 unidades en las ventas.
Gasto en Publicidad en Periódico (0.0043):

Por cada dólar adicional gastado en publicidad en Periódico, se espera un aumento de aproximadamente 0.0043 unidades en las ventas.

-----------------------------------------------------------------------

Interpretation
TV Advertising Spending (0.0545):

For each additional dollar spent on TV advertising, an increase of approximately 0.0545 units in sales is expected.
Radio Advertising Spending (0.1009):

For each additional dollar spent on radio advertising, an increase of approximately 0.1009 units in sales is expected.
Newspaper Advertising Spending (0.0043):

For each additional dollar spent on newspaper advertising, an increase of approximately 0.0043 units in sales is expected.

In [14]:

scores = cross_val_score(model, x, y, cv=5)  # 5-fold cross-validation

print ("Result:",scores)
print("Mean: ", np.mean(scores))
print("Std. Dev.: ", np.std(scores))
print("95% CI: ", np.percentile(scores, [2.5,97.5]))

Result: [0.87556263 0.93177791 0.92150403 0.84554586 0.90247132]
Mean:  0.8953723525274103
Std. Dev.:  0.03141250965862173
95% CI:  [0.84854754 0.93075052]


# Validacion Cruzada:

 En este caso, el modelo tiene un rendimiento bastante consistente en los diferentes pliegues, ya que las puntuaciones no varían significativamente y el intervalo de confianza es estrecho.

Un promedio de aproximadamente 0.8954 sugiere que, en promedio, el modelo explica alrededor del 89.54% de la variabilidad en los datos de respuesta. Esto es consistente con el alto R² que obtuve en la evaluación anterior.

La desviación estándar de aproximadamente 0.0314 indica que hay una pequeña variabilidad en los resultados entre los pliegues.

El intervalo de confianza del 95% proporciona una estimación de la variabilidad de los resultados y sugiere que se puede tener confianza en que el rendimiento del modelo está dentro de este rango.

-------------------------------------------------------------------------

Cross-Validation:
In this case, the model performs quite consistently across different folds, as the scores do not vary significantly, and the confidence interval is narrow.

An average of approximately 0.8954 suggests that, on average, the model explains around 89.54% of the variability in the response data. This is consistent with the high R² obtained in the previous evaluation.

The standard deviation of approximately 0.0314 indicates that there is little variability in the results between folds.

The 95% confidence interval provides an estimate of the variability of the results and suggests that there can be confidence in the model's performance falling within this range.

# Prediccion 
Ahora ingresemos valores en el modelo de acuerdo con las características que hemos utilizado para entrenarlo y predizcamos cuántas unidades del producto se pueden vender en función de la cantidad gastada en publicidad en varias plataformas:

In [15]:
feature= np.array([[230.1, 37.8,69.2]])
print(model.predict(feature))

[21.37254028]


 # Conclusión:
El modelo de regresión lineal parece ser bastante efectivo para predecir las ventas en función del gasto en publicidad en TV, Radio y Periódico. La validación cruzada respalda la robustez del modelo, y las evaluaciones indican que tiene un buen poder predictivo. Sin embargo, ten en cuenta que estos resultados se basan en las características específicas de tus datos y en la suposición de que la relación entre las variables es lineal. Siempre es recomendable realizar más análisis y considerar posibles mejoras en el modelo.

1. Juega un poco con las predicciones de este modelo!
2. No te olvides de seguirme
3. Dedique varias horas, puedes llevarte este trabajo. Solo dale un "Me Gusta"
4. Tus comentarios son valiosos, para mi! 

*Fuente original https://thecleverprogrammer.com/2022/03/01/future-sales-prediction-with-machine-learning/

ChatGPT me ayudo a redactar las explicaciones del proyecto*


-----------------------------------------------------------------------
Conclusion:
The linear regression model appears to be quite effective in predicting sales based on advertising spending on TV, Radio, and Newspaper. Cross-validation supports the model's robustness, and evaluations indicate it has good predictive power. However, keep in mind that these results are based on the specific characteristics of your data and the assumption that the relationship between variables is linear. It's always advisable to conduct further analysis and consider possible improvements to the model.

Play around with the predictions of this model!
Don't forget to follow me.
I dedicated several hours to this, so feel free to appreciate it with a "Like."
Your feedback is valuable to me!
*Original source: https://thecleverprogrammer.com/2022/03/01/future-sales-prediction-with-machine-learning/

ChatGPT helped me draft the project explanations.*
